scripts/Diffraction/isis_powder/routines/absorb_corrections.py

```diff
@@ -112,3 +114,47 @@ def _do_cylinder_absorb_corrections(ws_to_correct, multiple_scattering, is_vanad
     if previous_units != ws_units.tof:
         ws_to_correct = mantid.ConvertUnits(InputWorkspace=ws_to_correct, Target=previous_units)
     return ws_to_correct
+
+
+def apply_paalmanpings_absorb_and_subtract_empty(workspace, instrument, summed_empty, sample_details, run_number):
+    expected_filename = "PaalmanPingsCorrections_" + instrument.get_instrument_prefix() + str(run_number) \
+                        + "_" + instrument._get_input_batching_mode() + ".nxs"
+    save_directory = instrument._inst_settings.output_dir
+    force_recalculate_paalman_pings = instrument._inst_settings.force_recalculate_paalman_pings
+    corrections_file_found = False
+    try:
+        if expected_filename in os.listdir(save_directory):
+            corrections_file_found = True
+    except FileNotFoundError:
+        pass
+
+    if force_recalculate_paalman_pings or not corrections_file_found:
+        events_per_point_string = instrument._inst_settings.paalman_pings_events_per_point
+        container_geometry = sample_details.generate_container_geometry()
+        container_material = sample_details.generate_container_material()
+        if container_geometry and container_material:
+            mantid.SetSample(workspace, Geometry=sample_details.generate_sample_geometry(),
+                             Material=sample_details.generate_sample_material(),
+                             ContainerGeometry=container_geometry,
+                             ContainerMaterial=container_material)
+        else:
+            mantid.SetSample(workspace, Geometry=sample_details.generate_sample_geometry(),
+                             Material=sample_details.generate_sample_material())
+        if events_per_point_string:
+            events_per_point = int(events_per_point_string)
+        else:
+            events_per_point = 1000
+        corrections = mantid.PaalmanPingsMonteCarloAbsorption(InputWorkspace=workspace, Shape='Preset',
+                                                              EventsPerPoint=events_per_point)
+        mantid.SaveNexus(corrections, os.path.join(save_directory, expected_filename))
+    else:
+        expected_path = os.path.join(save_directory, expected_filename)
+        logger.warning(f'Loading pre-calculated corrections file found at {expected_path} \n'
+                       f' To recalculate with new sample + container values,'
+                       f' set "force_recalculate_paalman_pings = True" in the config file.')
+        corrections = mantid.LoadNexus(expected_path)
```

Comment:
I get an error when running a PP calculation where the corrections file already exists:

![image](https://user-images.githubusercontent.com/56295817/167398067-a6af723f-b156-44b6-ae87-8915327e8664.png)
The error comes from the ApplyPaalmanPingsCorrection call but I think it originates from the fact that this call to LoadNexus returns a tuple rather than a workspace group. This seems to be a feature of the C++ code for LoadNexus where it sets up multiple output workspace properties if the .nxs file is a group. The tuple has 5 things in it called OutputWorkspace, OutputWorkspace_1, OutputWorkspace_2 etc and the first one is the workspace group. So you probably want to set corrections equal to corrections.OutputWorkspace
Since the code path using the "cache" will be quite a common one I think it's worth having a test that goes through this bit of code as well. Fine to just run the calculation twice in the test